### PR TITLE
👉👆🏻👇Side choosing ability

### PR DESCRIPTION
**Pull request by stason314**
## Description
Simple function for setting view with `UIEdgeInsets` and ability to choosing sides which will be pinned to superview. For example, it's useful when you have insets const but need to use only two of them for some view.

## TODO
- [x] Cover by tests
- [x] Update readme

## Checklist

- [x] Follows code style
- [x] Has no SwiftLint warnings
- [x] Covered by tests
- [x] All tests are passed
- [x] Access control words were checked

### CODE REVIEW

File: Sources/Maker.swift [hunk 1]
```diff
@@ -565,6 +574,34 @@ public final class Maker {
...
+            case .bottom:
+                bottom(inset: insets.bottom)
+            case .left:
+                left(inset: insets.left)
+            case .right:
+                right(inset: insets.right)
+            case .top:
+                top(inset: insets.top)
+            case .horizontal:
+                top(inset: insets.top).bottom(inset: insets.bottom)
```
Comment (artemnovichkov) [inline Sources/Maker.swift]: I guess horizontal is related to left/right insets

Comment (stason314) [reply Sources/Maker.swift]: 🙈

File: Sources/Maker.swift [hunk 2]
```diff
@@ -24,6 +24,15 @@ public var nui_safeArea: SafeArea {
     return SafeArea()
 }
 
+public enum Sides {
```
Comment (Antowkos) [inline Sources/Maker.swift]: I suppose we need to add docs to this enum. For example when we should use it and how. Also we need explanation for each case in it

**Review by Antowkos — CHANGES_REQUESTED** [review]

File: Tests/MakerTests.swift [hunk 3]
```diff
@@ -361,6 +361,16 @@ class MakerTests: BaseTest {
...
+
+    func testThatCorrectlyConfiguresSliceOf_edge_insets_toSuperview() {
+
+        let insets: UIEdgeInsets = .init(top: 20, left: 10, bottom: 15, right: 10)
+        testingView.configureFrame { maker in
+            maker.edges(insets: insets, sides: .top, .left)
+            maker.size(width: 40, height: 40)
+        }
+        XCTAssertEqual(testingView.frame, CGRect(x: 10, y: 20, width: 40, height: 40))
+    }
```
Comment (artemnovichkov) [inline Tests/MakerTests.swift:437]: This API is not fully covered by this test. You should test ALL cases.

**Review by artemnovichkov — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -126,6 +126,15 @@ view.configureFrame { maker in
...
 the second method has optional parameters, so ```maker.edges(top: 5, left: 5, bottom: 5)``` also works correct, but does not create ```right``` relation, that in some cases is very useful.
 
+Also if you have `UIEdgeInsets` property and you want to use cup of them you can use
+
+```swift
+let insets: UIEdgeInsets = .init(top: 15, left: 10, bottom: 20, right: 10)
+view.configureFrame { maker in
+    maker.edges(insets: insets, sides: [.left, .bottom])
+}
+```
```
Comment (Antowkos) [inline README.md:142]: It would be nice to use `vertical` or/and `horizontal` sides in example as a more convenient way than `[.left, .right]`

File: Tests/MakerTests.swift [hunk 5]
```diff
@@ -382,7 +438,7 @@ class MakerTests: BaseTest {
         containet.addSubview(view2)
         
         containet.configureFrame { maker in
```
Comment (Antowkos) [inline Tests/MakerTests.swift]: Is `containet` a correct name of this variable? Or it is a typo?

**Review by Antowkos — CHANGES_REQUESTED** [review]

File: Sources/Maker.swift [hunk 6]
```diff
@@ -565,6 +577,34 @@ public final class Maker {
...
+    /// Creates edge relations for superview.
+    ///
+    /// - parameter insets: The insets for setting relations for superview.
+    ///
+    /// - parameter sides: The sides which will inculed from edge insets to setting relations.
+    ///
+    /// - returns: `Maker` instance for chaining relations.
+
+    @discardableResult public func edges(insets: UIEdgeInsets, sides: Set<Sides>) -> Maker {
+        sides.forEach { side in
```
Comment (artemnovichkov) [inline Sources/Maker.swift:575]: What's about removing old `edges` function and add `Sides.all` case and use it by default? Like:
```
@discardableResult 
public func edges(insets: UIEdgeInsets, sides: Set<Sides> = [.all]) -> Maker
```

File: Tests/MakerTests.swift [hunk 7]
```diff
@@ -361,7 +362,62 @@ class MakerTests: BaseTest {
...
+                }
+                else {
+                    x = mainView.frame.width - insets.right - testingView.frame.width
+                }
+            case .vertical:
+                y = insets.top
+                height = mainView.frame.height - insets.top - insets.bottom
+            }
+
+            XCTAssertEqual(testingView.frame, CGRect(x: x, y: y, width: width, height: height))
```
Comment (artemnovichkov) [inline Tests/MakerTests.swift]: It works better I guess, but it's hard to read and understand...

File: Tests/MakerTests.swift (hunk 5 above)
Comment (stason314) [reply Tests/MakerTests.swift]: 🤔Think there's should be `contained`

Comment (stason314) [reply Tests/MakerTests.swift]: Or better `containerView`

File: README.md [hunk 8]
```diff
@@ -126,6 +126,21 @@ view.configureFrame { maker in
...
+```swift
+let insets: UIEdgeInsets = .init(top: 15, left: 10, bottom: 20, right: 10)
+view.configureFrame { maker in
+    maker.edges(insets: insets, sides: [.left, .right])
+}
+```
+or
+```swift
+view.configureFrame { maker in
+maker.edges(insets: insets, sides: .horizontal)
```
Comment (artemnovichkov) [inline README.md]:
```suggestion
    maker.edges(insets: insets, sides: .horizontal)
```

File: Tests/ExtensionsTests.swift [hunk 9]
```diff
@@ -0,0 +1,67 @@
+//
+//  ExtensionsTests.swift
+//  FramezillaTests
+//
+//  Created by Стас Клюхин on 09/07/2019.
+//
+
+import XCTest
+
+class ExtensionsTests: BaseTest {
```
Comment (artemnovichkov) [inline Tests/ExtensionsTests.swift]:
```suggestion
final class ExtensionsTests: BaseTest {
```

File: Sources/Maker.swift [hunk 10]
```diff
@@ -24,6 +24,48 @@ public var nui_safeArea: SafeArea {
...
+    public static let horizontal: Sides = [.left, .right]
+
+    public static let all: Sides = [.vertical, .horizontal]
+    
+    public init(rawValue: Int) {
+        self.rawValue = rawValue
+    }
+    
+    public var description: String {
+        switch self {
```
Comment (artemnovichkov) [inline Sources/Maker.swift]: Why did you add the description?

File: Tests/ExtensionsTests.swift [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            })
+            switch options {
+            case .all:
+                XCTAssertEqual(iterationsCount, allOptionsCount)
+            case .option1:
+                XCTAssertEqual(iterationsCount, 1)
+            case .combinedOption:
+                XCTAssertEqual(iterationsCount, combinedOptionsCount)
+            default:
+                XCTAssertNil(nil)
```
Comment (artemnovichkov) [inline Tests/ExtensionsTests.swift]: Please add a messages for all asserts

**Review by artemnovichkov — CHANGES_REQUESTED** [review]

File: Sources/Maker.swift (hunk 10 above)
Comment (stason314) [reply Sources/Maker.swift]: Just for debug. I'll remove this

**Review by artemnovichkov — APPROVED** [review]

File: Sources/Maker.swift [hunk 12]
```diff
@@ -24,16 +24,25 @@ public var nui_safeArea: SafeArea {
     return SafeArea()
 }
 
-/// Sides enum.
+/// Sides option set.
```
Comment (Antowkos) [inline Sources/Maker.swift]: It seems like this line is redundant

File: Sources/Maker.swift [hunk 13]
```diff
@@ -1016,3 +1001,16 @@ public final class Maker {
         }
     }
 }
+
+extension OptionSet where RawValue: BinaryInteger, Element == Self  {
+    public func forEach(_ body: (Self) -> Void) {
```
Comment (Antowkos) [inline Sources/Maker.swift]: What is the purpose of this `extension`?

File: Tests/ExtensionsTests.swift [hunk 14]
```diff
@@ -0,0 +1,67 @@
+//
+//  ExtensionsTests.swift
+//  FramezillaTests
+//
+//  Created by Стас Клюхин on 09/07/2019.
```
Comment (Antowkos) [inline Tests/ExtensionsTests.swift]: Wrong headers. @artemnovichkov, could you help with default header for our example project?

**Review by Antowkos — CHANGES_REQUESTED** [review]

**Review by ReshetovE — APPROVED** [review]

**Review by ntunin — APPROVED** [review]

File: Tests/ExtensionsTests.swift (hunk 14 above)
Comment (artemnovichkov) [reply Tests/ExtensionsTests.swift]: Sure, let's fix it!

File: Sources/Maker.swift (hunk 13 above)
Comment (stason314) [reply Sources/Maker.swift]: Without this extension I can't use switch in `edges` function and should use `if` chaining. Are you sure its better than elegant `switch` in cycle or how I can do it better? + I use it in tests

**Review by Antowkos — CHANGES_REQUESTED** [review]
@stason314, please, double check your global .gitignore file and add `.DS_Store` files to it

**Review by Antowkos — APPROVED** [review]
